roxie/roxiemem/roxiemem.cpp

```diff
@@ -2922,14 +2941,93 @@ class BufferedRowCallbackManager
         Semaphore doneSem;
         struct
         {
+            unsigned slaveId;
             bool critical;
             bool result;
         } args;
         bool abort;
     };
 
+    class CallbackItem : public CInterface
+    {
+        typedef std::pair<unsigned, IBufferedRowCallback *> CallbackPair;
+        typedef std::vector<CallbackPair> CallbackPairArray;
+
+    public:
+        CallbackItem(unsigned _cost, unsigned _activityId) : cost(_cost), activityId(_activityId), nextCallback(0)
+        {
+        }
+        ~CallbackItem()
+        {
+        }
+        void addCallback(unsigned slaveId, IBufferedRowCallback * callback)
+        {
+            callbacks.push_back(CallbackPair(slaveId, callback));
+        }
+        inline bool isEmpty() const
+        {
+            return callbacks.empty();
+        }
+        inline bool matches(unsigned searchCost, unsigned searchActivityId) const
+        {
+            return (cost == searchCost) && (activityId == searchActivityId);
+        }
+        void removeCallback(unsigned searchSlaveId, IBufferedRowCallback * searchCallback)
+        {
+            for (CallbackPairArray::iterator iter = callbacks.begin(); iter != callbacks.end(); ++iter)
+            {
+                unsigned curSlave = iter->first;
+                if ((iter->first == searchSlaveId) && (iter->second == searchCallback))
+                {
+                    callbacks.erase(iter);
+                    return;
+                }
+            }
+            throwUnexpectedX("Removing callback that has not been registered");
+        }
+        unsigned releaseRows(BufferedRowCallbackManager & manager, unsigned whichSlave, unsigned minSuccess, bool critical)
+        {
+            CallbackPairArray::const_iterator begin = callbacks.begin();
+            CallbackPairArray::const_iterator end = callbacks.end();
+            CallbackPairArray::const_iterator iter = begin;
+            //MORE: Should nextCallback be tracked per slave?  There are arguments for and against.
+            if (nextCallback < callbacks.size())
+                iter += nextCallback;
```

Comment:
Is this safe? Could it be out of bounds if num callbacks has shrunk between calls?